File: parser/parser.go
package parser

import (
	"fmt"
	"goterp/scanner"
	"os"
)

type Parser struct {
	Tok scanner.Token
	S   scanner.Scanner
}

func Make(p *Parser, filepath string) {
	scanner.Make(&p.S, filepath)
	p.advance()
}

func (p *Parser) advance() {
	var tok scanner.Token

	p.S.Scan(&tok)
	p.Tok = tok
}

func (p *Parser) consume(kind scanner.TokenKind) {
	if p.Tok.Kind != kind {
		fmt.Fprintf(os.Stderr, "%s:%d expected %s but got %s\n", p.S.Filepath, p.Tok.Line, kind, p.Tok.Kind)
		os.Exit(1)
	}

	p.advance()
}

func (p *Parser) next(kinds ...scanner.TokenKind) bool {
	for _, k := range kinds {
		if p.Tok.Kind == k {
			return true
		}
	}
	return false
}

func (p *Parser) primary() Expr {
	if p.next(scanner.TokenIdent, scanner.TokenNum, scanner.TokenTrue, scanner.TokenFalse, scanner.TokenNil, scanner.TokenStr) {
		e := ExprLit{p.Tok}
		p.advance()
		return e
	}

	if p.next(scanner.TokenLParen) {
		p.advance()
		e := p.expression()
		p.consume(scanner.TokenRParen)
		return e
	}

	panic(fmt.Sprintf("%s:%d unknown primary %s", p.S.Filepath, p.Tok.Line, p.Tok.Kind))
}

func (p *Parser) unary() Expr {
	if p.next(scanner.TokenBang, scanner.TokenMinus) {
		op := p.Tok.Kind
		p.advance()
		return ExprUnary{op, p.unary()}
	}

	return p.primary()
}

func (p *Parser) factor() Expr {
	e := p.unary()

	if p.next(scanner.TokenStar, scanner.TokenSlash) {
		op := p.Tok.Kind
		p.advance()
		return ExprBinary{e, op, p.unary()}
	}

	return e
}

func (p *Parser) term() Expr {
	e := p.factor()

	if p.next(scanner.TokenPlus, scanner.TokenMinus) {
		op := p.Tok.Kind
		p.advance()
		return ExprBinary{e, op, p.factor()}
	}

	return e
}

func (p *Parser) comparison() Expr {
	e := p.term()

	if p.next(scanner.TokenLess, scanner.TokenGreater, scanner.TokenLessEq, scanner.TokenGreaterEq) {
		op := p.Tok.Kind
		p.advance()
		return ExprBinary{e, op, p.term()}
	}

	return e
}

func (p *Parser) equality() Expr {
	e := p.comparison()

	if p.next(scanner.TokenEqEq, scanner.TokenBangEq) {
		op := p.Tok.Kind
		p.advance()
		return ExprBinary{e, op, p.comparison()}
	}

	return e
}

func (p *Parser) logicAnd() Expr {
	e := p.equality()

	if p.next(scanner.TokenAnd) {
		p.advance()
		return ExprBinary{e, scanner.TokenAnd, p.equality()}
	}

	return e
}

func (p *Parser) logicOr() Expr {
	e := p.logicAnd()

	if p.next(scanner.TokenOr) {
		p.advance()
		return ExprBinary{e, scanner.TokenOr, p.logicAnd()}
	}

	return e
}

func (p *Parser) expression() Expr {
	return p.logicOr()
}

func (p *Parser) printStmt() Stmt {
	p.consume(scanner.TokenPrint)
	s := StmtPrint{p.expression()}
	p.consume(scanner.TokenSemicolon)

	return s
}

func (p *Parser) letStmt() Stmt {
	p.consume(scanner.TokenLet)

	name := p.Tok.Lit

	p.consume(scanner.TokenIdent)
	p.consume(scanner.TokenEq)

	e := p.expression()

	p.consume(scanner.TokenSemicolon)

	return StmtLet{name, e}
}

func (p *Parser) blockStmt() Stmt {
	var s StmtBlock

	p.consume(scanner.TokenLBrace)

	for !p.next(scanner.TokenRBrace) {
		s.Stmts = append(s.Stmts, p.statement())
	}

	p.consume(scanner.TokenRBrace)
	return s
}

func (p *Parser) assignStmt() Stmt {
	var s StmtAssign

	s.Name = p.Tok.Lit

	p.consume(scanner.TokenIdent)
	p.consume(scanner.TokenEq)

	s.Value = p.expression()

	p.consume(scanner.TokenSemicolon)
	return s
}

func (p *Parser) whileStmt() Stmt {
	p.consume(scanner.TokenWhile)
	return StmtWhile{p.expression(), p.blockStmt()}
}

func (p *Parser) statement() Stmt {
	if p.next(scanner.TokenPrint) {
		return p.printStmt()
	}

	if p.next(scanner.TokenLet) {
		return p.letStmt()
	}

	if p.next(scanner.TokenWhile) {
		return p.whileStmt()
	}

	if p.next(scanner.TokenIdent) {
		return p.assignStmt()
	}

	if p.next(scanner.TokenLBrace) {
		return p.blockStmt()
	}

	panic(fmt.Sprintf("%s:%d unknown statement %s", p.S.Filepath, p.Tok.Line, p.Tok.Kind))
}

func (p *Parser) Parse() []Stmt {
	ss := make([]Stmt, 0, 256)

	for p.Tok.Kind != scanner.TokenEof {
		ss = append(ss, p.statement())
	}

	return ss
}

File: parser.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "lib/mem.h"

#include "scanner.h"
#include "parser.h"

static void advance(struct parser *p)
{
    scan(&p->s, &p->tok);
}

static void consume(struct parser *p, enum token_kind kind)
{
    if (p->tok.kind != kind) {
        fprintf(stderr, "%s:%d:expected %s but got %s\n", p->s.filepath, p->tok.line, token_kind_str(kind), token_kind_str(p->tok.kind));
        exit(1);
    }

    advance(p);
}

void make_parser(struct parser *p, char *filepath)
{
    make_scanner(&p->s, filepath);
    advance(p);
}

static void primary(struct parser *p, struct expr *e)
{
    if (p->tok.kind == TOK_NUM) {
        struct expr_lit *el = malloc(sizeof(struct expr_lit));

        if (el == NULL) {
            perror("primary");
            exit(1);
        }

        memcpy(&el->value, &p->tok, sizeof(struct token));
        e->body = el;

        advance(p);
        return;
    }

    fprintf(stderr, "%s:%d:unknown primary expression %s\n", p->s.filepath, p->tok.line, token_kind_str(p->tok.kind));
    exit(1);
}

static void expression(struct parser *p, struct expr *e)
{
    primary(p, e);
}

static void print_stmt(struct parser *p, struct stmt *s)
{
    struct stmt_print *sp = malloc(sizeof(struct stmt_print));

    if (sp == NULL) {
        perror("print_stmt");
        exit(1);
    }

    advance(p);
    expression(p, &sp->value);
    consume(p, TOK_SEMICOLON);

    s->body = sp;
}

static void statement(struct parser *p, struct stmt *s)
{
    if (p->tok.kind == TOK_PRINT) {
        print_stmt(p, s);
        return;
    }

    fprintf(stderr, "%s:%d:unknown statement %s\n", p->s.filepath, p->tok.line, token_kind_str(p->tok.kind));
    exit(1);
}

void parse(struct parser *p, void *stmts)
{
    while (p->tok.kind != TOK_EOF) {
        struct stmt *s = memnext(stmts);
        statement(p, s);
    }
}
